### DISCUSSION TITLE
With the new comments, how do I remap them to one key?

### DISCUSSION THREAD

**Discussion by leet0rz** [neovim/neovim#29075] [Q&A]
```lua
remap("n", "'", ":norm gcc<CR>")
```

This first one seems to work for one liners, but I cant seem to get the block one to work as a remap no matter what I try.
So far I've tried these:

```lua
remap("v", "'", ":norm gvgc<CR>")
remap("v", "'", ":norm gcc<CR>")
```

I saw elsewhere that it is already a remap, but maybe there is a way for me to call the function that does the block commenting in visual mode, the 2 above mess up in different instances both of them.

**Comment by yochem** [comment]
This works perfectly fine for me:

```lua
vim.keymap.set("v", "'", ":norm gc<CR>")
```

Any reason you haven't tried that one yet?

Btw, generally using `<cmd>` instead of `:` in remaps is better. You could also opt for:

```lua
vim.keymap.set("v", "'", function()
  vim.cmd.norm('gc')
end)
```

;)

**Reply by neanias** [reply]
As a semi-related question, why is the `:norm` necessary? I've been trying to map using `vim.keymap.set("n", "<leader>cc", "gcc")` but it doesn't work. If I use `:/<cmd>norm`, it does work. Remapping things like `0` and `<C-w>j` work just fine, so I don't understand why this doesn't work...

**Reply by zeertzjq** [reply]
Add `remap = true` to `opts`

**Reply by neanias** [reply]
Ah, amazing, thank you!

**Comment by mangelozzi** [comment]
Loving this new built in command, let's me reduce plug in count. 
I found this worked quite well for me:

```
vim.keymap.set({"n", "v"}, "<leader>j", "gc", {remap = true})
vim.keymap.set({"n"}, "<leader>jj", "gcc", {remap = true})  -- Use with leader n to comment n lines
```

**Comment by leet0rz** [comment]
> This works perfectly fine for me:
> 
> ```lua
> vim.keymap.set("v", "'", ":norm gc<CR>")
> ```
> 
> Any reason you haven't tried that one yet?
> 
> Btw, generally using `<cmd>` instead of `:` in remaps is better. You could also opt for:
> 
> ```lua
> vim.keymap.set("v", "'", function()
>   vim.cmd.norm('gc')
> end)
> ```
> 
> ;)

the v one you have does not work, it needs to be just gc and have remap = true. These seem to be working for n and v:
```lua
remap("n", "'", ":norm gcc<CR>")
remap("v", "'", "gc", {remap = true})
```

**Comment by BlueDrink9** [comment]
As an alternative that doesn't rely on remaps (for example, if you use `gc` for something else), you can use the [underlying lua functions](https://github.com/neovim/neovim/blob/8df6736ca14d09f87cf0a8486758ac5708819434/runtime/lua/vim/_defaults.lua#L134) (although given they are namespaced with a `_` prefix, the interface might be liable to change? Not totally sure what the implications of that underscore are.)

```lua
local operator_rhs = function()
    return require('vim._comment').operator()
end
vim.keymap.set({ 'n', 'x' }, "'c", operator_rhs, { expr = true, desc = 'Toggle comment' })

local line_rhs = function()
    return require('vim._comment').operator() .. '_'
end
vim.keymap.set('n', "'cc", line_rhs, { expr = true, desc = 'Toggle comment line' })

local textobject_rhs = function()
    require('vim._comment').textobject()
end
vim.keymap.set({ 'o' }, "'c", textobject_rhs, { desc = 'Comment textobject' })
```

**Comment by looshch** [comment]
the solutions there don’t work if there is already a map for `gc`. Any ideas?

**Reply by BlueDrink9** [reply]
My suggestion above covers this case

**Reply by looshch** [reply]
hey @BlueDrink9, yep thanks for that! That’s what i currently [do](https://github.com/looshch/configs/blob/master/nvim.lua#L166), i was just wondering is there any way to not reach out to neovim’s internal api and instead do it in the traditional way, here are some [examples](https://github.com/looshch/configs/blob/master/nvim.lua#L25-L27) of what i mean by that

> although given they are namespaced with a `_` prefix, the interface might be liable to change

i didn’t even think about it but you are raising a very good point

**Reply by dundargoc** [reply]
This sounds not optimal. I don't have enough about the internals to know if there's a better approach. @echasnovski any ideas?

**Reply by echasnovski** [reply]
> That’s what i currently [do](https://github.com/looshch/configs/blob/master/nvim.lua#L166), i was just wondering is there any way to not reach out to neovim’s internal api and instead do it in the traditional way, here are some [examples](https://github.com/looshch/configs/blob/master/nvim.lua#L25-L27) of what i mean by that

I am afraid your two requirements (don't use `gc` yet do traditional remap without using functions) contradict each other. Neovim's keys for commenting are implemented through custom `gc` / `gcc` mappings and in order to do the plain remap those keys should be not remapped by the user.

If you don't want to rely on internal `require('vim._comment')` functions (which is a totally reasonable choice), you can get the target callback programmatically and reassign to different keys. With something like that:

```lua
local copy_keymap = function(mode, cur_lhs, new_lhs)
  local map_data = vim.fn.maparg(cur_lhs, mode, false, true)
  map_data.lhs, map_data.lhsraw = new_lhs, vim.keycode(new_lhs)
  vim.fn.mapset(map_data)
end

copy_keymap('n', 'gc', '<Leader>gc')
copy_keymap('x', 'gc', '<Leader>gc')
copy_keymap('o', 'gc', '<Leader>gc')
copy_keymap('n', 'gcc', '<Leader>gcc')
```

Make sure that this is executed while copied map data still points to built-in commenting.

**Reply by brianhuster** [reply]
> I am afraid your two requirements (don't use gc yet do traditional remap without using functions) contradict each other. 

Do you have plans to make these `gc` mappings core commands, similar to `gw`, `gq`, so on?

**Comment by vianneychin** [comment]
```lua
vim.keymap.set("v", "<leader>/", function()
	local count = vim.v.count
	vim.cmd.norm((count > 0 and count or "") .. "gcc")
end)

vim.keymap.set("n", "<leader>/", function()
	local count = vim.v.count
	vim.cmd.norm((count > 0 and count or "") .. "gcc")
end)

vim.keymap.set("o", "<leader>/", function()
	local count = vim.v.count
	vim.cmd.norm((count > 0 and count or "") .. "gcc")
end)

vim.keymap.set("x", "<leader>/", function()
	local count = vim.v.count
	vim.cmd.norm((count > 0 and count or "") .. "gcc")
end)
```

This is my current setup. Also works with the count behavior that normally works with `gcc`. e.g., `25gcc` would comment down 25 lines.

**Reply by RVPLiu** [reply]
why not this:
```
vim.keymap.set({"v", "n", "o", "x"}, "<leader>/", function()
	local count = vim.v.count
	vim.cmd.norm((count > 0 and count or "") .. "gcc")
end)
````

**Reply by yochem** [reply]
`vim.v.count1` could also come in handy here:

```
vim.cmd.norm(vim.v.count1 .. 'gcc')
```
